python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1121,342 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
+                P, form_compiler_options=form_compiler_options, jit_options=jit_options
+            )
+        )
+        self._a = _create_form(
+            J, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+
+        self.bcs = bcs
+        self.u = u
+
+    def F(self, snes: PETSc.SNES, x: PETSc.Vec, F: PETSc.Vec):
+        """Assemble the residual into the vector F.
+
+        Args:
+            snes: The solver instance
+            x: The vector containing the latest solution
+            F: Vector to assemble the residual into
+        """
+        assert x.getType() != "nest"
+        assert F.getType() != "nest"
+        with F.localForm() as f_local:
+            f_local.set(0.0)
+
+        self.replace_solution(x)
+        assemble_vector_block(F, self.L, self.a, bcs=self.bcs, x0=x, alpha=-1.0)
+
+    def J(self, snes, x, J, P):
+        """Assemble the Jacobian matrix.
+
+        Args:
+            snes: The solver instance
+            x: The vector containing the latest solution
+            J: Matrix to assemble the Jacobian into
+            P: Matrix to assemble the preconditioner into
+        """
+        self.replace_solution(x)
+        assert x.getType() != "nest" and J.getType() != "nest" and P.getType() != "nest"
+        J.zeroEntries()
+        assemble_matrix_block(J, self.a, bcs=self.bcs, diagonal=1.0)
+        J.assemble()
+        if self._a_prec is not None:
+            P.zeroEntries()
+            assemble_matrix_block(P, self._a_prec, bcs=self.bcs, diagonal=1.0)
+            P.assemble()
+
+    def replace_solution(self, x: PETSc.Vec):
+        """Update the solution for the unknown `u` with the values in `x`.
+
+        Args:
+            x: Data that is inserted into the solution vector
+        """
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+        offset_start = 0
+        for ui in self.u:
+            Vi = ui.function_space
+            num_sub_dofs = Vi.dofmap.index_map.size_local * Vi.dofmap.index_map_bs
+            ui.x.petsc_vec.array_w[:num_sub_dofs] = x.array_r[
+                offset_start : offset_start + num_sub_dofs
+            ]
+            ui.x.petsc_vec.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+            offset_start += num_sub_dofs
+
+    def copy_solution(self, x: PETSc.Vec):
+        """Copy the data in `u` into the vector `x`.
+
+        Args:
+            x: Vector to insert data into
+        """
+        u_petsc = [ui.x.petsc_vec.array_r for ui in self.u]
+        index_maps = [
+            (ui.function_space.dofmap.index_map, ui.function_space.dofmap.index_map_bs)
+            for ui in self.u
+        ]
+        _cpp.la.petsc.scatter_local_vectors(
+            x,
+            u_petsc,
+            index_maps,
+        )
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+
+
+class NestSNESProblem(BlockSNESProblem):
```

Comment:
I see your point with repeated constructor, but in that case I think a cleaner solution is to refactor the common logic into a free function that does the Jacobian AD and call it in both problem types. Otherwise `class NestSNESProblem(BlockSNESProblem)` is not the right relation,  nest problem is not a special type of block problem.

Reply:
@michalhabera I've now removed the inheritance, and instead introduced a `typing.Protocol` that illustrates what a class that is passed into `SNESSolver` expects.